bot/exts/valentines/markov_poem_generator.py

```diff
@@ -0,0 +1,513 @@
+        traceback: Optional[TracebackType]
+    ):
+        # If an error is raised, then the word may not be added to the cache.
+        if self.word in self.cache and len(self.cache[self.word]) == 0:
+            logging.info(f"No rhymes were found for the word: {self.word}")
+
+
+def memoize(func: Callable) -> Callable:
+    """
+    A decorator to access and cache rhyme sets.
+
+    If the rhyme set of the word already exists in the cache, then
+    there is no need to execute `func` because results can be taken from the
+    cache. Otherwise, execute `func` and store the results into the cache.
+    """
+    @functools.wraps(func)
+    async def wrapper(
+        *args,
+        instance: Callable = None,
+        word: str = None
+    ) -> Set[str]:
+        with Cache(word) as (cache, word):
+            if word not in cache.keys():
+                cache[word] = await func(instance, word)
+
+        return cache[word]
+
+    return wrapper
+
+
+def command_wrapper(func: Callable) -> Callable:
+    """The decorator adds a timeout and shows the bot's `typing...` text."""
+    @functools.wraps(func)
+    async def wrapper(
+        *args,
+        instance: Callable = None,
+        timeout: int = None,
+        ctx: commands.Context = None,
+        scheme: str = None,
+        unit_count: Dict[str, int] = None,
+        time_start: datetime.time = None,
+        is_first_error: bool = None
+    ) -> Callable:
+        try:
+            async with async_timeout.timeout(timeout), ctx.typing():
+                return await func(instance, ctx, scheme, unit_count,
+                                  time_start, is_first_error)
+        except TimeoutError:
+            await ctx.send(f"Sorry {ctx.author.mention}, but your poem"
+                           " timed out, please try again!")
+
+    return wrapper
+
+
+class MarkovPoemGenerator(commands.Cog):
+    """
+    A cog that provides a poem by taking the markov of a corpus.
+
+    By processing the corpus text through a markov chain, a series of lines
+    can be iterated through, whilst corresponding to the given rhyme scheme.
+    """
+
+    POEM_TIMEOUT = 30  # In seconds
+    near_rhyme_min_score = 2000
+    rhyming_line_finder_limiter = 20000
+    max_char_range = (50, 120)  # For the sentence generator
+
+    SOURCES: List[str] = [
+        "shakespeare_corpus.txt"
+    ]
+
+    templates: Dict[str, str] = {
+        "shakespearean-sonnet": "abab/cdcd/efef/gg",
+        "spenserian-sonnet": "abab/bcbc/cdcd/ee",
+        "petrarch-sonnet": "abbaabba/cdecde",
+        "ballade": "ababbcbc",
+        "terza-rima": "aba/bcb/cdc/ded/ee",
+        "villanelle": "aba/aba/aba/aba/aba/abaa",
+        "limerick": "aabba"
+    }
+
+    rhyme_websites: List[Tuple[bool, str]] = [
+        # (is exact rhyme, website link)
+        (True, "https://api.datamuse.com/words?rel_rhy="),
+        (False, "https://api.datamuse.com/words?rel_nry=")
+    ]
+
+    def __init__(self, bot: commands.Bot):
```

Comment:
We should type hint `bot.bot.Bot` here and in the `setup` function

Reply:
But `markov_poem` isn't supposed to return anything. I am not sure if I understand.